### DISCUSSION TITLE
Smooth fonts above pictures

### DISCUSSION THREAD

**Discussion by DamirAS2020** [Bodmer/TFT_eSPI#970] [Q&A - Fonts]
Hi,

I do some characters menu above a picture.
Color of strings needs to be changed (not changing a picture).
After that characters is become not smooth ( both colours).
To exclude my wrong things I tried to change your example
Smooth_font_reading_TFT

  if (st) tft.setTextColor(TFT_BLACK, TFT_WHITE); // Background color is ignored if callback is set
  else tft.setTextColor(TFT_RED, TFT_WHITE);
  st = !st;

I moved 
  rainbow_fill();
from loop to setup.

bool st = true;

is above  loop.

Both colors have become not smooth.

How to solve the problem?

**Comment by Bodmer** [comment]
First you need to check the TFT can be read. This is quite simple, use drawPixel, then readPixel and check the colours match.  Not all displays can be read. Once that is checked then the example "Smooth_font_reading_TFT" will work correctly.

**Comment by Bodmer** [comment]
Depending on which display driver you have (assuming SPI) then you may need to reduce the read SPI clock frequency, there is an option to do this in the setup file.

**Reply by tanmanh0707** [reply]
Hello @Bodmer ,

I'm not sure if it is ok to post comment to this very old topic but I have the same problem in displaying the TFT ST3755 128x128
Whenever I used setTextColor, the font becomes not smooth as before. Below is my simple code, the customfont.h is a header file contains any font array from example folder. I tested with all font array examples and all of them have the same issue.

```
#include <TFT_eSPI.h>
#include "customfont.h"

#define SCREEN_WIDTH    128
#define SCREEN_HEIGHT   128

TFT_eSPI    tft = TFT_eSPI();

void screen_puts_tc(uint16_t y, char *str)
{
  tft.setTextDatum(TC_DATUM);
  tft.setTextColor(0xFFFF);  //This line causes font broken (not smooth)
  tft.setTextWrap(true);
  tft.drawString(str, SCREEN_WIDTH/2, y);
}

bool initTFT()
{
  tft.init();
  tft.fillScreen(TFT_RED);
  tft.loadFont(customfont12);
  screen_puts_tc(48, "Scanning...");
  return true;
}

void setup()
{
  Serial.begin(115200);
  initTFT();
}

void loop(void){}
```

Could you please take a look at it and help me with this issue?

Thank you very much
Manh

**Reply by Bodmer** [reply]
The background colour must be specified, in this case use:
tft.setTextColor(0xFFFF,  TFT_RED);

**Reply by tanmanh0707** [reply]
Hello Bodmer,

Thank you for your time.
The above code is just an example of my source code. Actually I have an image and want to draw text on it such as a highlight bar image with text.
The background color can be whatever and not solid. Therefore, to specify background color is very difficult. Is there another way to do that?

Best regards, 
Manh

**Reply by Bodmer** [reply]
See Smooth_font_reading_TFT example.

That example requires a display that allows pixel colours to be read back.

**Reply by tanmanh0707** [reply]
Hi Bodmer,

My display is ST7735 which does not support reading pixel color back. However, even so, the background image color is not solid, it is kind of gradient color. So If I can read the pixel color, do I need to set multiple background color in setTextColor?

Thank you,
Manh

**Comment by Bodmer** [comment]
In that case the Smooth_font_gradient is more appropriate.  If background is not specified the background pixel colour will be requested via the callback.

**Reply by tanmanh0707** [reply]
Dear Bodmer,

Thank you very much for your time and your great contribution. Hope you have a nice day.
